onnx/defs/nn/defs.cc

```diff
@@ -23,6 +23,94 @@ static std::string auto_pad_doc =
 } // namespace ONNX_NAMESPACE
 
 namespace ONNX_NAMESPACE {
+
+void convPoolTypeAndShapeInference(InferenceContext& ctx) {
+  propagateElemTypeFromInputToOutput(ctx, 0, 0);
+
+  if (!hasNInputShapes(ctx, 2)) {
+    return;
+  }
+
+  // don't bother with legacy auto_pad for now
+  if (ctx.getAttribute("auto_pad")) {
+    return;
+  }
+
+  size_t n_input_dims = (size_t) (ctx.getInputType(0)->tensor_type().shape().dim_size() - 2);
+
+  std::vector<int64_t> dilations;
+  if (getRepeatedAttribute(ctx, "dilations", dilations)) {
```

Comment:
I don't see attribute "dilations" mentioned in these ops ... should it be added to these ops in spec?

Reply:
the pooling ops don't have dilations, but Conv does, and it's harmless to share the code since it is a no-op when unspecified.

However I will add a comment to make this clear in the code

Reply:
It's not harmless for models where the pooling ops specify a dilations attribute. The pooling ops implementation will ignore the dilations if specified (unless some attribute-checking throws out the model as an invalid model), but the shape inference will take them into account and produce a wrong shape, leading to possible memory corruption if memory is allocation on this basis.

Reply:
May be we could add a Boolean parameter to the inference function to indicate if dilations should be considered or not?

Reply:
do we allow arbitrary extra attributes (e.g. putting a dilation on something that doesn't support it?) If so then yes this is a problem.

In any case yes a flag sounds like a good idea